### DISCUSSION TITLE
Adding Nvidia Variable Rate Shading to Filament

### DISCUSSION THREAD

**Discussion by yavuz650** [google/filament#6541] [Q&A] [answered]
Nvidia GPUs support Variable Rate Shading(VRS). Nvidia introduced new functions and enumerations to OpenGL, Vulkan and DirectX for developers to integrate VRS into their applications. I want to add this feature to Filament in my forked repo. 

The problem is, VRS needs to be enabled only for the main render pass, and it should be turned off for other passes such as shadow map rendering.

I'm using OpenGL so I think I just need to modify the OpenGL backend and incorporate the necessary API calls to enable VRS. 

Any tips on which classes/files I should focus on?

**Comment by bejado** [comment, marked as answer]
A good place to start might be by adding a new field to [`RenderPassParams`](https://github.com/google/filament/blob/1653d3615b79b5fd0b41df7fba727e6a5ee7e66a/filament/backend/include/backend/DriverEnums.h#L1018), then setting this field for the appropriate passes. For example, in the color pass we set a bunch of fields on `out.params` depending on the configuration: https://github.com/google/filament/blob/1653d3615b79b5fd0b41df7fba727e6a5ee7e66a/filament/src/RendererUtils.cpp#L201-L210.

The OpenGL driver can then read this field inside of `OpenGLDriver::beginRenderPass` and make the appropriate calls.

**Reply by yavuz650** [reply]
I see, thank you for the tip!

How should the user enable/disable VRS? I'm thinking maybe I could add a new enable variable to `filament::FView`. Then, read this variable in the color pass to set the field in `RenderPassParams`, which will be passed down to `OpenGLDriver::beginRenderPass`. This interface would be consistent with Filament's existing features, too, no?

**Reply by bejado** [reply]
Yes, a new API on `View` would probably be how we would approach it.
